### PR TITLE
Configurable call fee refund for signed submissions

### PR DESCRIPTION
**Pull request by emostov**
Previously, the winning solution had it deposit unreserved, and was rewarded with the reward base + call fee. All other unchecked solutions had their deposit unreserved, but still lost their call fee. This PR changes things so a runtime can be configured to refund the call fees for some (or all) of the stored, non-winning solutions.

Configuring `SignedMaxRefunds`:

The first `x` unchecked solutions can get their call fee refunded by setting `SignedMaxRefunds` to `x`. This will lead to no ejected solutions getting a refund.

polkadot companion: https://github.com/paritytech/polkadot/pull/5309

### CODE REVIEW

**Comment by kianenigma** [comment]
if not overly complicated, can we make it such that you specify how many of the "non-winnings" get a refund? i.e 'type MaxFullRefunds'. This way we can enable and merge this on Kusama ahead of time as well.

File: frame/election-provider-multi-phase/src/signed.rs [hunk 1]
```diff
@@ -52,6 +52,8 @@ pub struct SignedSubmission<AccountId, Balance: HasCompact, Solution> {
 	pub raw_solution: RawSolution<Solution>,
 	/// The reward that should potentially be paid for this solution, if accepted.
 	pub reward: Balance,
+	// The estimated fee `who` paid to submit the solution.
+	pub call_fee: Balance,
```
Comment (emostov) [inline frame/election-provider-multi-phase/src/signed.rs:56]: `reward` was a balance, and `call_fee` is a balance, so this should decode the same through runtime upgrades.

File: frame/election-provider-multi-phase/src/signed.rs [hunk 2]
```diff
@@ -395,10 +395,20 @@ impl<T: Config> Pallet<T> {
 		// Any unprocessed solution is pointless to even consider. Feasible or malicious,
 		// they didn't end up being used. Unreserve the bonds.
 		let discarded = all_submissions.len();
-		for SignedSubmission { who, deposit, .. } in all_submissions.drain() {
+		let mut count = 0;
+		for SignedSubmission { who, deposit, call_fee, .. } in all_submissions.drain() {
+			if T::SignedMaxRefunds::get().map_or(true, |max| count < max) {
```
Comment (ggwpez) [inline frame/election-provider-multi-phase/src/signed.rs:400]: You are doing one map access here per loop, you can probably save that in a temp to lower the weight.

**Review by ggwpez — COMMENTED** [review]
Just some comments :smile:

File: frame/election-provider-multi-phase/src/lib.rs [hunk 3]
```diff
@@ -628,6 +628,12 @@ pub mod pallet {
 		#[pallet::constant]
 		type SignedMaxWeight: Get<Weight>;
 
+		/// The maximum amount of unchecked solutions to refund the call fee for. If `None`, all
+		/// solutions will get a refund. Should either be `None` or a value less than
+		/// [`Self::SignedMaxSubmissions`]
+		#[pallet::constant]
+		type SignedMaxRefunds: Get<Option<u32>>;
```
[the remark this answers is not in the document]
Comment (emostov) [reply frame/election-provider-multi-phase/src/lib.rs:635]: Yeah: https://github.com/paritytech/substrate/pull/11002#issuecomment-1063758396 - although not sure if I agree that this is useful

[the remark this answers is not in the document]
Comment (ggwpez) [reply frame/election-provider-multi-phase/src/lib.rs:635]: So we could keep it for not and create an issue to remove it later.

File: frame/election-provider-multi-phase/src/lib.rs [hunk 4]
```diff
@@ -628,6 +628,12 @@ pub mod pallet {
 		#[pallet::constant]
 		type SignedMaxWeight: Get<Weight>;
 
+		/// The maximum amount of unchecked solutions to refund the call fee for. If `None`, all
+		/// solutions will get a refund. Should either be `None` or a value less than
+		/// [`Self::SignedMaxSubmissions`]
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/lib.rs:633]: I don't see an `integrity_test` for this invariant 😀

File: frame/election-provider-multi-phase/src/lib.rs [hunk 5]
```diff
@@ -987,14 +993,17 @@ pub mod pallet {
...
+				T::EstimateCallFee::estimate_call_fee(&call, None.into())
 			};
 
-			let submission =
-				SignedSubmission { who: who.clone(), deposit, raw_solution: *raw_solution, reward };
+			let submission = SignedSubmission {
+				who: who.clone(),
+				deposit,
+				raw_solution: *raw_solution,
+				call_fee,
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/lib.rs:1005]: now I remember why we are storing the reward: currently it is constant. But initially we thought it could be variable based on the score that the solution is providing etc. 

I still think that it can be computed on the fly, so nonetheless swapping it out of the struct is 👍 for me.

File: frame/election-provider-multi-phase/src/signed.rs [hunk 6]
```diff
@@ -395,10 +395,20 @@ impl<T: Config> Pallet<T> {
 		// Any unprocessed solution is pointless to even consider. Feasible or malicious,
 		// they didn't end up being used. Unreserve the bonds.
 		let discarded = all_submissions.len();
-		for SignedSubmission { who, deposit, .. } in all_submissions.drain() {
+		let mut count = 0;
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/signed.rs:398]:
```suggestion
		let mut refunded = 0;
```

File: frame/election-provider-multi-phase/src/signed.rs [hunk 7]
```diff
@@ -395,10 +395,20 @@ impl<T: Config> Pallet<T> {
...
+		for SignedSubmission { who, deposit, call_fee, .. } in all_submissions.drain() {
+			if T::SignedMaxRefunds::get().map_or(true, |max| count < max) {
+				// Refund fee
+				let _ = T::Currency::deposit_creating(&who, call_fee);
+				count += 1;
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/signed.rs:403]:
```suggestion
				refunded += 1;
```

File: frame/election-provider-multi-phase/src/signed.rs (hunk 2 above)
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/signed.rs:400]:
```suggestion
			if T::SignedMaxRefunds::get().map_or(true, |max| refunded < max) {
```

File: frame/election-provider-multi-phase/src/signed.rs [hunk 8]
```diff
@@ -395,10 +395,20 @@ impl<T: Config> Pallet<T> {
...
+			}
+
+			// Unreserve deposit
 			let _remaining = T::Currency::unreserve(&who, deposit);
-			weight = weight.saturating_add(T::DbWeight::get().writes(1));
 			debug_assert!(_remaining.is_zero());
+
+			Self::finalize_signed_phase_handle_refund(&who, deposit, call_fee);
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/signed.rs:410]: you are doing a double `unreserve` here. Once in this function and once in the line above.

Comment (kianenigma) [reply frame/election-provider-multi-phase/src/signed.rs:410]: `finalize_signed_phase_handle_refund` is opaque in what it does IMO, I would call it `fn refund_deposit_and_fee`, this would have made the issue clearer.

**Review by kianenigma — COMMENTED** [review]
As discussed elsewhere, when we push out a solution in the middle of the signed phase we should also refund that potentially. 

For these solutions, I presume we always refund their call-fee? hard to decide on it.

File: frame/election-provider-multi-phase/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (emostov) [reply frame/election-provider-multi-phase/src/lib.rs:635]: I actually now think this is useful, it helps reduce the incentive for solution churn.

File: frame/election-provider-multi-phase/src/signed.rs [hunk 9]
```diff
@@ -631,19 +642,60 @@ mod tests {
...
+		ExtBuilder::default().build_and_execute(|| {
+			roll_to(15);
+			assert!(MultiPhase::current_phase().is_signed());
+			assert_eq!(SignedMaxRefunds::get(), Some(1));
+			assert!(SignedMaxSubmissions::get() > 2);
+
+			for s in 0..SignedMaxSubmissions::get() {
+				let account = 99 + s as u64;
+				Balances::make_free_balance_be(&account, 100);
+				// score is always decreasing better
```
Comment (emostov) [inline frame/election-provider-multi-phase/src/signed.rs:656]:
```suggestion
				// score is always decreasing
```

File: frame/election-provider-multi-phase/src/signed.rs [hunk 10]
```diff
@@ -631,19 +642,60 @@ mod tests {
...
+			assert!(MultiPhase::finalize_signed_phase());
+
+			for s in 0..SignedMaxSubmissions::get() {
+				let account = 99 + s as u64;
+				// lower accounts have higher scores
+				if s == 0 {
+					// winning solution always gets call fee + reward
+					assert_eq!(balances(&account), (100 + 8 + 7, 0))
+				} else if s == 1 {
+					// 1 runner up gets there call fee refunded
```
Comment (emostov) [inline frame/election-provider-multi-phase/src/signed.rs:673]:
```suggestion
					// 1 runner up gets their call fee refunded
```

File: frame/election-provider-multi-phase/src/lib.rs [hunk 11]
```diff
@@ -628,6 +628,11 @@ pub mod pallet {
 		#[pallet::constant]
 		type SignedMaxWeight: Get<Weight>;
 
+		/// The maximum amount of unchecked solutions to refund the call fee for. If `None`, all
+		/// solutions, including ejected solutions, will get a refund.
+		#[pallet::constant]
+		type SignedMaxRefunds: Get<Option<u32>>;
```
[the remark this answers is not in the document]
Comment (kianenigma) [reply frame/election-provider-multi-phase/src/lib.rs:634]: but yeah nevermind :D

File: frame/election-provider-multi-phase/src/lib.rs [hunk 12]
```diff
@@ -1013,6 +1021,10 @@ pub mod pallet {
 			let ejected_a_solution = maybe_removed.is_some();
 			// if we had to remove the weakest solution, unreserve its deposit
 			if let Some(removed) = maybe_removed {
+				if T::SignedMaxRefunds::get().is_none() {
+					// There are no limits on how many solutions get their call fee refunded
+					let _ = T::Currency::deposit_creating(&removed.who, removed.call_fee);
+				}
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/lib.rs:1027]:
```suggestion
```

File: frame/election-provider-multi-phase/src/lib.rs [hunk 13]
```diff
@@ -628,6 +628,11 @@ pub mod pallet {
 		#[pallet::constant]
 		type SignedMaxWeight: Get<Weight>;
 
+		/// The maximum amount of unchecked solutions to refund the call fee for. If `None`, all
+		/// solutions, including ejected solutions, will get a refund.
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/lib.rs:632]:
```suggestion
		/// The maximum amount of unchecked solutions to refund the call fee for.
```

Comment (kianenigma) [reply frame/election-provider-multi-phase/src/lib.rs:632]: TODO: integrity check to make sure this does not go above `T::SignedMaxSubmissions` (although it is harmless, but it clearly shows whoever is configuring the pallet that they've no idea what they are doing)

**Review by kianenigma — COMMENTED** [review]
Few final comments.

File: frame/election-provider-multi-phase/src/signed.rs [hunk 14]
```diff
@@ -395,10 +395,19 @@ impl<T: Config> Pallet<T> {
 		// Any unprocessed solution is pointless to even consider. Feasible or malicious,
 		// they didn't end up being used. Unreserve the bonds.
 		let discarded = all_submissions.len();
-		for SignedSubmission { who, deposit, .. } in all_submissions.drain() {
+		let mut refund_count = 0;
+		let max_refunds = T::SignedMaxRefunds::get();
+		for SignedSubmission { who, deposit, call_fee, .. } in all_submissions.drain() {
+			if refund_count < max_refunds {
+				// Refund fee
+				let _ = T::Currency::deposit_creating(&who, call_fee);
```
[the remark this answers is not in the document]
Comment (emostov) [reply frame/election-provider-multi-phase/src/signed.rs:403]: Yeah, both are technically minting funds. Despite the name, call fee refund is minting new funds since the actual fee probably went to block producer/treasury. So I think your suggestion makes sense

[the remark this answers is not in the document]
Comment (emostov) [reply frame/election-provider-multi-phase/src/signed.rs]: 4867e2feef6dd011c24a76863b9b4c3d1dce0f23

File: frame/election-provider-multi-phase/src/lib.rs [hunk 15]
```diff
@@ -986,14 +995,17 @@ pub mod pallet {
 
 			// create the submission
 			let deposit = Self::deposit_for(&raw_solution, size);
-			let reward = {
+			let call_fee = {
 				let call = Call::submit { raw_solution: raw_solution.clone() };
-				let call_fee = T::EstimateCallFee::estimate_call_fee(&call, None.into());
-				T::SignedRewardBase::get().saturating_add(call_fee)
```
Comment (niklasad1) [inline frame/election-provider-multi-phase/src/lib.rs:992]: dq: why is this removed?

**Review by niklasad1 — COMMENTED** [review]
looks clean Zeke but I have a few dumb questions...

File: bin/node/runtime/src/lib.rs [hunk 16]
```diff
@@ -600,7 +600,8 @@ frame_election_provider_support::generate_solution_type!(
 );
 
 parameter_types! {
-	pub MaxNominations: u32 = <NposSolution16 as frame_election_provider_support::NposSolution>::LIMIT as u32;
+	pub MaxNominations: u32 = <NposSolution16 as sp_npos_elections::NposSolution>::LIMIT as u32;
```
[the remark this answers is not in the document]
Comment (emostov) [reply bin/node/runtime/src/lib.rs]: 9f8a136baa861cbb3a4480294c64012556efa239

File: frame/election-provider-multi-phase/src/lib.rs (hunk 15 above)
Comment (emostov) [reply frame/election-provider-multi-phase/src/lib.rs:992]: We used to calculate the reward here, but now we just want the call fee. This is because we now calculate the reward on the spot when paying it out, but we can't calculate the call fee when paying it out because it is highly variable.

The reward is `T::SignedRewardBase::get().saturating_add(call_fee)`

On the other hand, to get the call fee we just need `T::EstimateCallFee::estimate_call_fee(&call, None.into())`.

**Comment by emostov** [comment]
Changes in https://github.com/paritytech/substrate/pull/11002/commits/07d8c9d50cf5b8f92d04e134b8504167313b8d44 should address https://github.com/paritytech/srlabs_findings/issues/192

**Review by kianenigma — APPROVED** [review]
Still LGTM

File: frame/election-provider-multi-phase/src/signed.rs [hunk 17]
```diff
@@ -235,20 +235,34 @@ impl<T: Config> SignedSubmissions<T> {
...
+		keys.sort();
+
 		SignedSubmissionIndices::<T>::kill();
 		SignedSubmissionNextIndex::<T>::kill();
-		let insertion_overlay = sp_std::mem::take(&mut self.insertion_overlay);
-		SignedSubmissionsMap::<T>::drain()
-			.filter(move |(k, _v)| !self.deletion_overlay.contains(k))
-			.map(|(_k, v)| v)
-			.chain(insertion_overlay.into_iter().map(|(_k, v)| v))
+		let _ = sp_std::mem::take(&mut self.deletion_overlay);
```
Comment (emostov) [inline frame/election-provider-multi-phase/src/signed.rs:261]:
```suggestion
```

[the remark this answers is not in the document]
Comment (emostov) [reply frame/election-provider-multi-phase/src/signed.rs]: [1cf7847](https://github.com/paritytech/substrate/pull/11002/commits/1cf7847d2f9cf458b2c26acdd2be6ce45c946bc0)

File: frame/election-provider-multi-phase/src/signed.rs [hunk 18]
```diff
@@ -235,20 +236,33 @@ impl<T: Config> SignedSubmissions<T> {
...
+			.filter(|k| {
+				if self.deletion_overlay.contains(k) {
+					// Remove submissions that should be deleted.
+					SignedSubmissionsMap::<T>::remove(k);
+					false
+				} else {
+					true
+				}
+			})
+			.chain(self.insertion_overlay.keys().map(|n| *n))
```
Comment (niklasad1) [inline frame/election-provider-multi-phase/src/signed.rs:256]:
```suggestion
			.chain(self.insertion_overlay.keys().copied()
```

File: frame/election-provider-multi-phase/src/signed.rs [hunk 19]
```diff
@@ -235,20 +236,33 @@ impl<T: Config> SignedSubmissions<T> {
...
 	/// from the underlying `SignedSubmissionsMap`, putting the storages into an invalid state.
 	///
 	/// Note that, like `put`, this function consumes `Self` and modifies storage.
-	fn drain(mut self) -> impl Iterator<Item = SignedSubmissionOf<T>> {
+	fn drain_submitted_order(mut self) -> impl Iterator<Item = SignedSubmissionOf<T>> {
+		let mut keys = SignedSubmissionsMap::<T>::iter_keys()
+			.filter(|k| {
+				if self.deletion_overlay.contains(k) {
+					// Remove submissions that should be deleted.
+					SignedSubmissionsMap::<T>::remove(k);
```
Comment (niklasad1) [inline frame/election-provider-multi-phase/src/signed.rs:250]: I think `SignedSubmissionsMap::<T>::remove(k)` (side-effects) in filter is a bit unexpected...

Personally I would prefer `filter_map` or something even if the actual iterator isn't affected by it.

**Review by niklasad1 — APPROVED** [review]

File: frame/election-provider-multi-phase/src/signed.rs [hunk 20]
```diff
@@ -235,20 +236,33 @@ impl<T: Config> SignedSubmissions<T> {
...
+					false
+				} else {
+					true
+				}
+			})
+			.chain(self.insertion_overlay.keys().copied()
```
Comment (niklasad1) [inline frame/election-provider-multi-phase/src/signed.rs:256]:
```suggestion
			.chain(self.insertion_overlay.keys().copied())
```

**Comment by kianenigma** [comment]
bot merge

**Comment by kianenigma** [comment]
bot merge

**Comment by emostov** [comment]
bot merge

**Comment by emostov** [comment]
bot merge
